File: Python/102. Binary Tree Level Order Traversal.py
"""
Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).
"""
# 102. Binary Tree Level Order Traversal
from typing import Optional, List
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        
        result = []
        queue = deque([root])
        
        while queue:
            level_size = len(queue)
            level_nodes = []

            for _ in range(level_size):
                node = queue.popleft()
                level_nodes.append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(level_nodes)

        return result
        
    
# Test Code
solution = Solution()
root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7))) # Expect [[3],[9,20],[15,7]]
#root = TreeNode(1) # Expect [[1]]
#root = TreeNode(None) # Expect []
print(solution.levelOrder(root))

File: Javascript/102. Binary Tree Level Order Traversal.js
/*
Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).
*/
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}
/**
 * 102. Binary Tree Level Order Traversal
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    if (root === null) { return []; }
    const result = [];
    
    const dfs = function(node, level) {
        if (node === null) { return; }
        if (!result[level]) result[level] = [];
        result[level].push(node.val);
        dfs(node.left, level + 1);
        dfs(node.right, level + 1);
    };
    
    dfs(root, 0);
    return result;
};
const root = new TreeNode(3,
    new TreeNode(9, null, null),
    new TreeNode(20,
        new TreeNode(15, null, null),
        new TreeNode(7, null, null)
    )
);// Expect [[3],[9,20],[15,7]]

//const root = new TreeNode(1, null, null); // Expect [[1]]
//const root = new TreeNode(null, null, null); // Expect []

console.log(levelOrder(root));
